File: src/algorithms/python/trie.py
# Trie (Prefix Tree) Insertion and Word Matching in Python
# Implements efficient character-by-character string retrieval.

class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the prefix tree.
        """
        node = self.root
        for char in word.upper():
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        """
        Returns true if the word is found in the Trie.
        """
        node = self.root
        for char in word.upper():
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end_of_word

    def starts_with(self, prefix: str) -> bool:
        """
        Returns true if there is any word in the Trie that starts with the given prefix.
        """
        node = self.root
        for char in prefix.upper():
            if char not in node.children:
                return False
            node = node.children[char]
        return True

# Initialize Trie and insert sample words
trie = Trie()
trie.insert("CAT")
trie.insert("CAR")
trie.insert("DOG")

print("Search 'CAT':", trie.search("CAT")) # True
print("Search 'CAN':", trie.search("CAN")) # False
print("Prefix 'CA':", trie.starts_with("CA")) # True

File: src/algorithms/cpp/trie.cpp
// Trie (Prefix Tree) Insertion and Word Matching in C++
// Implements efficient character-by-character string retrieval.

#include <iostream>
#include <unordered_map>
#include <string>

struct TrieNode {
    std::unordered_map<char, TrieNode*> children;
    bool isEndOfWord = false;
};

class Trie {
private:
    TrieNode* root;

public:
    Trie() {
        root = new TrieNode();
    }

    void insert(std::string word) {
        TrieNode* node = root;
        for (char c : word) {
            c = toupper(c);
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isEndOfWord = true;
    }

    bool search(std::string word) {
        TrieNode* node = root;
        for (char c : word) {
            c = toupper(c);
            if (node->children.find(c) == node->children.end()) {
                return false;
            }
            node = node->children[c];
        }
        return node->isEndOfWord;
    }

    bool startsWith(std::string prefix) {
        TrieNode* node = root;
        for (char c : prefix) {
            c = toupper(c);
            if (node->children.find(c) == node->children.end()) {
                return false;
            }
            node = node->children[c];
        }
        return true;
    }
};

int main() {
    Trie trie;
    trie.insert("CAT");
    trie.insert("CAR");
    trie.insert("DOG");

    std::cout << "Trie initialized and words inserted successfully." << std::endl;
    return 0;
}
